### PR TITLE
Untested: Fix #1780 /Optimizations for Commshare gadget

### LINKED ISSUE #1780: Fix Transfering to commshared teams (teams with size=0)
Commshare gadget needs to be updated to transfer units from commshared players to the team that they're on.

### PR DESCRIPTION
**Pull request by ArchShaman**
- Fixes #1780
- Localized all LUA/Spring API stuff.
- Prevent griefing using malicious widgets (Sending overly long strings to commshare could lead to lag)
- Got rid of poor table usage. (Such as ProccessCommand's table)

This should make it less resource intensive.

### CODE REVIEW

File: LuaRules/Gadgets/game_share_mode.lua [hunk 1]
```diff
@@ -10,47 +10,79 @@ function gadget:GetInfo()
...
+local gMatch = string.gmatch
+local gSub = string.gsub
+local Find = string.find
+local Lowercase = string.lower
+local tonum = tonumber
+local tostr = tostring
+local Type = type
+local Select = select
+-- Spring API --
+local Echo = Spring.Echo
```
Comment (GoogleFrog) [inline LuaRules/Gadgets/game_share_mode.lua]: If you're going to localise Spring.Word then it must be spWord.

File: LuaRules/Gadgets/game_share_mode.lua [hunk 2]
```diff
@@ -10,47 +10,79 @@ function gadget:GetInfo()
...
+local public = {public = true}
+local Invites = {}
+local controlledplayers = {} -- table containing which team a playerid should be under.
+local controlledteams = {} -- contains which team a team of players should be under.
+local originalteamids = {} -- takes playerid as the key, gives the team as the value.
+local originalunits = {} -- contains which units are owned by a team that has commshared.
+local GaiaID = -9999
 
+------------------ speedups ------------------
+local gMatch = string.gmatch
```
Comment (GoogleFrog) [inline LuaRules/Gadgets/game_share_mode.lua]: Don't do any of this, it hurts readability.

File: LuaRules/Gadgets/game_share_mode.lua [hunk 3]
```diff
@@ -10,47 +10,79 @@ function gadget:GetInfo()
...
-	local strtbl = {}
-	for w in string.gmatch(str, "%S+") do
-	strtbl[#strtbl+1] = w
-	end
-	return strtbl
+if gadgetHandler:IsSyncedCode() == false then
+	return
 end
 
 local modOptions = {}
```
Comment (GoogleFrog) [inline LuaRules/Gadgets/game_share_mode.lua]: Why do you make an empty table then do nothing with it?

File: LuaRules/Gadgets/game_share_mode.lua [hunk 4]
```diff
@@ -10,47 +10,79 @@ function gadget:GetInfo()
...
 }
--- check config --
-if config.mergetype == nil then config.mergetype = "invite"; end
+modOptions = nil
+if config.mergetype == nil then config.mergetype = "invite"; end -- Set the config up in case. This line controls the default state (in case "sharemode" doesn't exist for some reason.)
+if config.mergetype == "all" then config.unmerging = false else config.unmerging = true end -- disable unmerging for all mode, but allow it for invite only
 
-if config.mergetype == "all" then config.unmerging = false else config.unmerging = true end
+------------------ Variables ------------------
+local private = {private = true}
```
Comment (GoogleFrog) [inline LuaRules/Gadgets/game_share_mode.lua]: The first two entries here are constants.

File: LuaRules/Gadgets/game_share_mode.lua (hunk 3 above)
Comment (ArchShaman) [inline LuaRules/Gadgets/game_share_mode.lua]: It's used here, mainly to set the config and turn off gadget if the sharemode == "off". However it's not used again past this point.

File: LuaRules/Gadgets/game_share_mode.lua (hunk 2 above)
Comment (ArchShaman) [inline LuaRules/Gadgets/game_share_mode.lua]: Don't localize lua functions? Ok.

File: LuaRules/Gadgets/game_share_mode.lua (hunk 1 above)
Comment (ArchShaman) [inline LuaRules/Gadgets/game_share_mode.lua]: Done. Do you want a new pr with the changes?

**Pull request merged**